### PR TITLE
TickCounter class refactoring

### PR DESCRIPTION
**Pull request by alexandre-lecoq**

### CODE REVIEW

**Comment by maehne** [comment]
Thanks for the PR @alexandre-lecoq! Could you please provide a bit more information, what you changed and why to facilitate its review?

**Comment by alexandre-lecoq** [comment]
Hello,
While investigating why the tick rate was slow in simulations, i stumbled on this class and noticed that propagationCompleted() which is often called was doing lots of work, so I made propagationCompleted() lighter, and also removed the arrays, and tried to make the code easier to understand at first glance.

**Review by maehne — APPROVED** [review]
Thanks @alexandre-lecoq for the PR and explanation! Looks good to me, but I suggest that another one, who has better insight into the modified code, has a closer look before we merge this PR.

**Comment by maehne** [comment]
@davidhutchens and @kevinawalsh: Could you maybe have a look at this PR as well?

**Review by davidhutchens — CHANGES_REQUESTED** [review]
This PR does clean up the code, using a java Queue rather than building one.

Note that the PR only changes the displayed actual clock frequency (when active) and does not change the way the simulator deals with clock ticks or the actual speed of the simulation. I do not suggest the scope be expanded, I am just clarifying the scope of the PR.


Bugs:
If you run at high frequency and then change frequency to something very low (like .5Hz) the first displayed value is NaN.

If you stop the clock (Auto-click disabled) and then restart it, the clock rate drops a lot and slowly moves back up. Likely, the old ticks are counted and the duration of time for the stopped ticks shows as a reduced rate. Restarting the clock should reset the TickCounter (and the SmoothingFilter).

Specifically, places the old code called updateSimulator, you now just assign the simulator and do not clear the SmoothingFilter. That may lead to miscalculated frequencies.


Code Style:
Your new files use cr-lf line endings. Our files use the Unix standard lf without the cr. Please remove the cr character from the end of each line.

Our code base uses camel case without underscores for variable names. You have underscores to start the instance variables of your classes. Please remove that underscore. Note that your constants with all upper-case and imbedded underscores is consistent with our style.

Most of your local variables are not changed and could be declared final.

Please fix the bugs and clean up the code style.

**Comment by maehne** [comment]
Thanks for the thorough review @davidhutchens!

**Review by maehne — CHANGES_REQUESTED** [review]
@alexandre-lecoq: Please address the issues raised by @davidhutchens!

**Comment by alexandre-lecoq** [comment]
> If you run at high frequency and then change frequency to something very low (like .5Hz) the first displayed value is NaN.
Not reproduced, but I removed the smoothing filter, it could have been the cause for that.

> If you stop the clock (Auto-click disabled) and then restart it, the clock rate drops a lot and slowly moves back up. Likely, the old ticks are counted and the duration of time for the stopped ticks shows as a reduced rate. Restarting the clock should reset the TickCounter (and the SmoothingFilter).
Fixed

> Specifically, places the old code called updateSimulator, you now just assign the simulator and do not clear the SmoothingFilter. That may lead to miscalculated frequencies.
Fixed

> Your new files use cr-lf line endings. Our files use the Unix standard lf without the cr. Please remove the cr character from the end of each line.
Files removed. Basic smoothing was enough.

> Our code base uses camel case without underscores for variable names
Fixed

> Most of your local variables are not changed and could be declared final.
Fixed

Also, I moved to System.nanoTime() rather than System.currentTimeMillis(), since there's a tick frequency option at 2048 KHz which corresponds to a period of 488.3 nanoseconds (well under 1 millisecond), I suppose it won't hurt.

File: src/main/java/com/cburch/logisim/gui/main/TickCounter.java
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/gui/main/TickCounter.java:29]: I recommend using the number of seconds before weight reduction. You can calculate the ticksThreshold and weightReductionTicks from the number of seconds and the computed fullTicksPerSecond. A value of around 5 seconds worked in my playing with it.

The value you have is so large that you likely never drop earlier things. At 100 ticks per second, it takes a long time to get to a billion ticks. And you really want a larger number of ticks when the clock is running fast and a smaller number when it is ticking slow.

I have sample hardware simulations that can change the speed during a run because parts of the hardware shut down (i.e. don't change state). So I can test it on that. It should change the reported speed as the simulation changes how fast it is actually running.

**Review by davidhutchens — CHANGES_REQUESTED** [review]
I really like this idea for implementation. It is much simpler than what we have, or your first attempt to improve it. I made some specific comments in the code that I think will make it better.

By the way, you do not need to force-push an update. You can just add a commit to your branch when you address these suggestions.

File: src/main/java/com/cburch/logisim/gui/main/TickCounter.java
Comment (alexandre-lecoq) [reply src/main/java/com/cburch/logisim/gui/main/TickCounter.java]: Fixed

File: src/main/java/com/cburch/logisim/gui/main/TickCounter.java
Comment (alexandre-lecoq) [reply src/main/java/com/cburch/logisim/gui/main/TickCounter.java]: Fixed

File: src/main/java/com/cburch/logisim/gui/main/TickCounter.java
Comment (alexandre-lecoq) [reply src/main/java/com/cburch/logisim/gui/main/TickCounter.java:29]: It was a mistake I introduced when switching to nanoseconds.
I was supposed to be 1000.
Hopefully, it'll give a better feeling like that.

**Comment by alexandre-lecoq** [comment]
> By the way, you do not need to force-push an update. You can just add a commit to your branch when you address these suggestions.

Yes, I forgot again, I got new habits after starting using gerrit.

**Comment by R3dst0ne** [comment]
There are huge numbers in multiple places. Using Scientific notation might help with readability and decrease the chance of errors by typing the wrong amount of 0's.

**Review by MarcinOrlowski — CHANGES_REQUESTED** [review]
As per @R3dst0ne comment, please do change the code and replace `500000000` (multiple places) by less error-prone notation scientific notation `5E+8` (you can put unfolded value in comment if you really wish).

**Comment by alexandre-lecoq** [comment]
I removed all those big values completely and only kept a unique constant with thousands separators
`static final int NANOSECONDS_PER_SECONDS = 1_000_000_000;`
Does that address the issue ?
I'm not sure it really needs to be a double.

**Comment by R3dst0ne** [comment]
The unique constants are a good solution.
Still, I would prefer to use scientific notation in the definitions.

**Comment by MarcinOrlowski** [comment]
@R3dst0ne we already do use `000_000_000` notation so I'd leave it as it as this approach also helps.

**Review by davidhutchens — APPROVED** [review]
I think the integer constants with underscores is better than floating point constants in scientific notation for variables declared as integers. So I would leave this as is.

This implementation looks clean to me. I tried it with some LE files and it has a similar feel to the original.

Thank you @alexandre-lecoq for your contribution.

File: src/main/java/com/cburch/logisim/gui/main/TickCounter.java
Comment (MarcinOrlowski) [inline src/main/java/com/cburch/logisim/gui/main/TickCounter.java:28]: I'm wondering if `UNIT_UPDATE_THRESHOLD_IN_NANOSECONDS` is not better name for this const. Not crucial, yet `ns` thing is least important here so it should not be first word in the constant name (IMHO). Leaving final decision to you though.

File: src/main/java/com/cburch/logisim/gui/main/TickCounter.java
Comment (MarcinOrlowski) [inline src/main/java/com/cburch/logisim/gui/main/TickCounter.java:43]: Please add comment why `12`

**Review by MarcinOrlowski — CHANGES_REQUESTED** [review]

File: src/main/java/com/cburch/logisim/gui/main/TickCounter.java
Comment (alexandre-lecoq) [reply src/main/java/com/cburch/logisim/gui/main/TickCounter.java]: Done

File: src/main/java/com/cburch/logisim/gui/main/TickCounter.java
Comment (alexandre-lecoq) [reply src/main/java/com/cburch/logisim/gui/main/TickCounter.java]: Done

**Review by maehne — APPROVED** [review]
Looks good to me, but it should get reviewed and tested also by @BFH-ktt1 before merging this PR. Thanks for your effort in polishing this PR @alexandre-lecoq!

**Comment by BFH-ktt1** [comment]
I tested and it looks good to me, I would like to have this in V3.8.0. @MarcinOrlowski : could you deblock?

**Pull request merged**
